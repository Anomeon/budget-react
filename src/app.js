import React from 'react';
import ReactDOM from 'react-dom';
import {
  Router,
  Route,
  hashHistory
} from 'react-router';
import {
  EnterData,
  Menu,
  Report,
  Categories
} from './components';

ReactDOM.render(
  <Router history={hashHistory}>
    <Route component={Menu}>
      <Route path="/" component={EnterData}/>
      <Route path="/items/:category" component={EnterData}/>
      <Route path="test" component={Report}/>
      <Route path="categories" component={Categories}/>
    </Route>
  </Router>,
  document.getElementById('app')
);
