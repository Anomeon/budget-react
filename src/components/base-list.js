import React from 'react';

export class BaseList extends React.Component {
  render() {
    return (
      <ul>
        {
          this.props.items.map((item) => {
            return <li key={item.id}>{item[this.props.type]}</li>
          })
        }
      </ul>
    );
  }
};
