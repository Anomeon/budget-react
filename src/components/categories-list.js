import React from 'react';

export class CategoriesList extends React.Component {
  render() {
    return (
      <ul>
        {
          this.props.items.map((item) => {
            return <li key={item.id}>{item.category}</li>
          })
        }
      </ul>
    );
  }
};
