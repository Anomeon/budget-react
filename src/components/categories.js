import React from 'react';
import {BaseList} from './index';
import {ItemStorage} from '../services';

export class Categories extends React.Component {
  constructor() {
    super();
    this.storage = new ItemStorage(localStorage);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.state = {
      items: this.storage.getItems('categories', true)
    }
  }

  handleSubmit(e) {
    e.preventDefault();
    let formData = {};
    [...document.querySelectorAll('input')]
      .forEach(el => formData[el.name] = el.value);
    this.storage.addItem('categories', formData);
    e.target.reset();
    this.setState({items: this.storage.getItems('categories', true)});
  }

  render() {
    return (
      <div>
        <form onSubmit={this.handleSubmit}>
          <input type="text" name="category" placeholder="Enter category" autoFocus={true}/>
          <button>Add category</button>
        </form>
        <BaseList items={this.state.items} type="category"/>
      </div>
    );
  }
};
