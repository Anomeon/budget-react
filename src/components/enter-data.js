import React from 'react';
import {ItemStorage} from '../services/item-storage';
import {List} from './index';

let storage = new ItemStorage(localStorage);

export class EnterData extends React.Component {

  constructor(props) {
    super(props);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleClearAll = this.handleClearAll.bind(this);
    this.state = {
      items: storage.getItems('items', true)
    }
  }

  handleSubmit(e) {
    e.preventDefault();
    let formData = {};
    Array.prototype.slice.call(document.querySelectorAll('input'))
      .forEach(el => formData[el.name] = el.value);
    storage.addItem('items', formData);
    e.target.reset();
    this.setState({items: storage.getItems('items', true)})
  }

  handleClearAll() {
    storage.deleteItems(
      this.state.items.map((item) => {
        return item.id;
      })
    );
    this.setState({items: storage.getItems('items', true)})
  }

  render() {
    return (
      <div>
        <form onSubmit={this.handleSubmit}>
          <label htmlFor="item"> Item </label>
          <input name="item" type="string" id="item"/>
          <label htmlFor="sum"> Sum </label>
          <input name="sum" type="number" id= "sum"/>
          <button>Send</button>
          <button type="button" onClick={this.handleClearAll}>Clear All</button>
        </form>
        <List items={this.state.items}/>
      </div>
    );
  }
};
