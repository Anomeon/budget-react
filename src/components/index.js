export {EnterData} from './enter-data';
export {ListItem} from './list-item';
export {List} from './list';
export {Menu} from './menu';
export {Test} from './test';
export {Categories} from './categories';
export {CategoriesList} from './categories-list';
