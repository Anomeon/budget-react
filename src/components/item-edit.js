import React from 'react';
import {ItemStorage} from '../services';

export class ItemEdit extends React.Component {
  constructor() {
    super();
    this.storage = new ItemStorage(localStorage);
    this.categories = this.storage.getItems('categories', true);
    this.tags = this.storage.getItems('tags', true);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleDelete = this.handleDelete.bind(this);
  }

  componentWillMount() {
    this.itemId = this.props.params.id;
    this.item = this.storage.getItems('items')[this.itemId];
  }

  handleSubmit(e) {
    e.preventDefault();
    let formData = {};
    [...document.querySelectorAll('input')]
      .forEach(el => formData[el.name] = el.value);
    [...e.target.querySelectorAll('select:not([name=tags])')]
      .forEach(el => formData[el.name] = el.value);
    formData['tags'] = [...document.querySelector('[name=tags]').options]
      .filter(option => option.selected)
      .map(option => option.value);
    this.storage.updateItem('items', this.itemId, formData);
    window.location.hash = '#';
  }

  handleDelete(e) {
    this.storage.deleteItems('items', [this.itemId]);
    window.location.hash = '#';
  }

  render() {
    const {item, sum, category} = this.item;
    return (
      <form onSubmit={this.handleSubmit}>
        <input name="item" type="string" id="item" placeholder="Enter item" autoFocus={true} defaultValue={item}/>
        <input name="sum" type="number" id= "sum" placeholder="Enter sum" defaultValue={sum}/>
        <select name="category" defaultValue={category}>
          <option readOnly defaultValue label=" "></option>
          {
            this.categories.map(category => <option key={category.id}>{category.category}</option>)
          }
        </select>
        <select name="type">
          <option>expenses</option>
          <option>income</option>
        </select>
        <select name="tags" multiple>
          {
            this.tags.map(tag => <option key={tag.id}>{tag.tag}</option>)
          }
        </select>
        <button type="submit">Save</button>
        <button onClick={this.handleDelete} type="button">Delete</button>
      </form>
    );
  }
};
