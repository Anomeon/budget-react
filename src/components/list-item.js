import React from 'react';
import {Link} from 'react-router';

export class ListItem extends React.Component {
  render() {
    const {id, item, sum, category} = this.props.item;
    return (
      <div>
        <div>item: {item}</div>
        <div>sum: {Math.abs(sum)}</div>
        <div>category: {category}</div>
        <Link to={`items\/${id}\/edit`}>Edit</Link>
      </div>
    );
  }
};
