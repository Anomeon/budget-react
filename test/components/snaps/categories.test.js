import React from 'react';
import {Categories} from '../../../src/components';
import renderer from 'react-test-renderer';

it('renders correctly', () => {
  const tree = renderer.create(
    <Categories/>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
