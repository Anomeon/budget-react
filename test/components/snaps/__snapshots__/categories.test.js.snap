exports[`test renders correctly 1`] = `
<div>
  <form
    onSubmit={[Function]}>
    <input
      autoFocus={true}
      name="category"
      placeholder="Enter category"
      type="text" />
    <button>
      Add category
    </button>
  </form>
  <ul>
    <li>
      test
    </li>
  </ul>
</div>
`;
